### PR TITLE
Port katello_sync_plan to foreman_helper

### PR DESCRIPTION
**Pull request by mdellweg**

### CODE REVIEW

File: plugins/module_utils/foreman_helper.py [hunk 1]
```diff
@@ -131,6 +131,38 @@ def _patch_location_api(self):
         _location_update_params_location = next(x for x in _location_update['params'] if x['name'] == 'location')
         _location_update_params_location['params'].append(_location_organizations_parameter)
 
+    def _patch_sync_plan_api(self):
+        """This is a workaround for the broken sync_plan apidoc in katello.
```
Comment (evgeni) [inline plugins/module_utils/foreman_helper.py]: where is the Katello issue link? :)

Comment (mdellweg) [reply plugins/module_utils/foreman_helper.py]: https://projects.theforeman.org/issues/27532

**Comment by evgeni** [comment]
needs a rebase after #394

File: plugins/module_utils/foreman_helper.py [hunk 2]
```diff
@@ -132,6 +132,39 @@ def _patch_location_api(self):
...
+            _sync_plan_methods = self.foremanapi.apidoc['docs']['resources']['sync_plans']['methods']
+
+            _sync_plan_add_products = next(x for x in _sync_plan_methods if x['name'] == 'add_products')
+            if next((x for x in _sync_plan_add_products['params'] if x['name'] == 'organization_id'), None) is None:
+                _sync_plan_add_products['params'].append(_organization_parameter)
+
+            _sync_plan_remove_products = next(x for x in _sync_plan_methods if x['name'] == 'remove_products')
+            if next((x for x in _sync_plan_remove_products['params'] if x['name'] == 'organization_id'), None) is None:
+                _sync_plan_remove_products['params'].append(_organization_parameter)
+        except Exception:
```
Comment (evgeni) [inline plugins/module_utils/foreman_helper.py]: in #392 I patched that under KatelloApypieModule, as sync plans are a Katello thing.

File: plugins/modules/katello_sync_plan.py [hunk 3]
```diff
@@ -80,85 +80,70 @@
...
+        products = module.find_resources_by_name('products', products, params=scope, thin=True)
+        desired_product_ids = set(p['id'] for p in products)
+        current_product_ids = set(p['id'] for p in entity['products']) if entity else set()
 
         if desired_product_ids != current_product_ids:
             if not module.check_mode:
                 product_ids_to_add = desired_product_ids - current_product_ids
                 if product_ids_to_add:
-                    sync_plan.add_products(data={'product_ids': list(product_ids_to_add)})
+                    payload = scope.copy()
```
Comment (evgeni) [inline plugins/modules/katello_sync_plan.py]: I thought you prefer `payload.update(scope)`? ;-)

File: plugins/modules/katello_sync_plan.py [hunk 4]
```diff
@@ -80,85 +80,70 @@
...
+                    payload['id'] = sync_plan['id']
+                    payload['product_ids'] = list(product_ids_to_add)
+                    module._resource_action('sync_plans', 'add_products', payload)
```
Comment (evgeni) [inline plugins/modules/katello_sync_plan.py]:
```suggestion
                    module.resource_action('sync_plans', 'add_products', payload)
```

File: plugins/modules/katello_sync_plan.py [hunk 5]
```diff
@@ -80,85 +80,70 @@
...
                 product_ids_to_remove = current_product_ids - desired_product_ids
                 if product_ids_to_remove:
-                    sync_plan.remove_products(data={'product_ids': list(product_ids_to_remove)})
+                    payload = scope.copy()
+                    payload['id'] = sync_plan['id']
+                    payload['product_ids'] = list(product_ids_to_remove)
+                    module._resource_action('sync_plans', 'remove_products', payload)
```
Comment (evgeni) [inline plugins/modules/katello_sync_plan.py]:
```suggestion
                    module.resource_action('sync_plans', 'remove_products', payload)
```

File: plugins/modules/katello_sync_plan.py [hunk 6]
```diff
@@ -80,85 +80,70 @@
...
-    changed, sync_plan = naildown_entity(SyncPlan, entity_dict, sync_plan, state, module)
+    changed, sync_plan = module.ensure_entity('sync_plans', entity_dict, entity, params=scope)
 
-    if products is not None:
-        products = find_products(module, products, entity_dict['organization'])
-        desired_product_ids = set(p.id for p in products)
-        current_product_ids = set(p.id for p in sync_plan.product)
+    if not module.desired_absent and products is not None:
+        products = module.find_resources_by_name('products', products, params=scope, thin=True)
+        desired_product_ids = set(p['id'] for p in products)
```
Comment (evgeni) [inline plugins/modules/katello_sync_plan.py]:
```suggestion
        desired_product_ids = set(product['id'] for product in products)
```

File: plugins/modules/katello_sync_plan.py [hunk 7]
```diff
@@ -80,85 +80,70 @@
...
+        current_product_ids = set(p['id'] for p in entity['products']) if entity else set()
```
Comment (evgeni) [inline plugins/modules/katello_sync_plan.py]:
```suggestion
        current_product_ids = set(product['id'] for product in entity['products']) if entity else set()
```

File: plugins/modules/katello_sync_plan.py (hunk 3 above)
Comment (mdellweg) [reply plugins/modules/katello_sync_plan.py]: This must have been a day older.

File: plugins/modules/katello_sync_plan.py [hunk 8]
```diff
@@ -80,85 +80,70 @@
     organization: "Default Organization"
     interval: "weekly"
     enabled: false
-    sync_date: "2017/01/01 00:00:00 +0000"
+    sync_date: "2017-01-01 00:00:00 UTC"
```
Comment (evgeni) [inline plugins/modules/katello_sync_plan.py]: should the expected date format be documented somewhere?

Comment (mdellweg) [reply plugins/modules/katello_sync_plan.py]: This example is part of the documentation.
I think, katello understands other formats, too. It's just not idempotent then.

**Pull request merged**